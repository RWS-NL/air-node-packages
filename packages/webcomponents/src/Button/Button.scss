@import '../styles/vars';

.button {
  @extend %fontNormal;
  background-color: $skyBlue100;
  border-radius: 0;
  float: right;
  height: 3.2rem;
  line-height: initial;
  margin-right: 1%;
  text-transform: none;

  &Label {
    @extend %fontNormal;
    color: $white;
    font-size: 1rem;
    text-transform: none;
  }

  &Secondary {
    background-color: $yellow100;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }

  &Outlined {
    background-color: $white;
    color: $skyBlue100;

    > span {
      color: $skyBlue100;
    }

    &Secondary {
      background-color: $white;
      color: $textBlack;

      > span {
        color: $textBlack;
      }
    }
  }

  &Disabled {
    background-color: $grey35 !important;
    border-color: $grey35 !important;
    box-shadow: none !important;
  }
}
