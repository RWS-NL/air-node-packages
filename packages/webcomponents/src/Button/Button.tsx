import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import css from './Button.scss';
import classnames from 'classnames';
import React, { FC, ReactNode, memo } from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 */
export const Button: FC<MUIButtonProps<'button', ButtonProps>> = props => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
};

export default memo(Button);
