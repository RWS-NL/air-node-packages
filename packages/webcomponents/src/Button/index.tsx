import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import classnames from 'classnames';
import React, { memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */
export const Button = memo((props: MUIButtonProps<'button', ButtonProps>) => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
});

export default Button;
