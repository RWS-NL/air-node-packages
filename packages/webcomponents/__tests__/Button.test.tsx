import { Button } from '@src/Button';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

const mockButtonClick = jest.fn();

describe('Component Tests', () => {
  let button: ShallowWrapper;

  beforeAll(() => {
    button = shallow(<Button variant='contained' color='primary' label='testButton' onClick={mockButtonClick} />);
  });

  test('button function called onClick', () => {
    button.simulate('click');
    expect(mockButtonClick).toHaveBeenCalledWith();
    expect(mockButtonClick).toHaveBeenCalledTimes(1);
  });

  test('button contains the correct translation', () => {
    expect(button.children().text()).toBe('testButton');
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const button = shallow(<Button variant='contained' color='primary' label='test' onClick={mockButtonClick} />);
    expect(button).toMatchSnapshot();
  });

  test('Varying Props', () => {
    const button = shallow(
      <Button variant='outlined' color='secondary' label={<div>snapshot label</div>} onClick={mockButtonClick} />
    );
    expect(button).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const button = shallow(
      <Button
        variant='outlined'
        color='secondary'
        label={<div>snapshot label</div>}
        onClick={mockButtonClick}
        data-qa='button-qa'
        customclasses={['sample-class']}
        disabled
      />
    );
    expect(button).toMatchSnapshot();
  });
});
