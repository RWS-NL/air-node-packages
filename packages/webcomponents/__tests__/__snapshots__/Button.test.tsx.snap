// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Button))
  className="sample-class"
  classes={
    Object {
      "containedPrimary": "buttonPrimary makeStyles-buttonShadow-1",
      "containedSecondary": "buttonSecondary makeStyles-buttonShadow-1",
      "disabled": "buttonDisabled",
      "label": "buttonLabel",
      "outlined": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedPrimary": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedSecondary": "buttonOutlinedSecondary makeStyles-buttonShadow-1",
      "root": "button",
    }
  }
  color="secondary"
  customclasses={
    Array [
      "sample-class",
    ]
  }
  data-qa="button-qa"
  disabled={true}
  label={
    <div>
      snapshot label
    </div>
  }
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  variant="outlined"
>
  <div>
    snapshot label
  </div>
</WithStyles(ForwardRef(Button))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Button))
  className=""
  classes={
    Object {
      "containedPrimary": "buttonPrimary makeStyles-buttonShadow-1",
      "containedSecondary": "buttonSecondary makeStyles-buttonShadow-1",
      "disabled": "buttonDisabled",
      "label": "buttonLabel",
      "outlined": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedPrimary": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedSecondary": "buttonOutlinedSecondary makeStyles-buttonShadow-1",
      "root": "button",
    }
  }
  color="primary"
  label="test"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  variant="contained"
>
  test
</WithStyles(ForwardRef(Button))>
`;

exports[`Snapshot Testing Varying Props 1`] = `
<WithStyles(ForwardRef(Button))
  className=""
  classes={
    Object {
      "containedPrimary": "buttonPrimary makeStyles-buttonShadow-1",
      "containedSecondary": "buttonSecondary makeStyles-buttonShadow-1",
      "disabled": "buttonDisabled",
      "label": "buttonLabel",
      "outlined": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedPrimary": "buttonOutlined makeStyles-buttonShadow-1",
      "outlinedSecondary": "buttonOutlinedSecondary makeStyles-buttonShadow-1",
      "root": "button",
    }
  }
  color="secondary"
  label={
    <div>
      snapshot label
    </div>
  }
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  variant="outlined"
>
  <div>
    snapshot label
  </div>
</WithStyles(ForwardRef(Button))>
`;
